src/main/kotlin/com/terraformation/backend/species/db/SpeciesStore.kt

```diff
@@ -478,65 +482,71 @@ class SpeciesStore(
-            .from(SPECIES_ECOSYSTEM_TYPES)
-            .where(SPECIES_ECOSYSTEM_TYPES.SPECIES_ID.eq(speciesId))
-            .fetch(SPECIES_ECOSYSTEM_TYPES.ECOSYSTEM_TYPE_ID.asNonNullable())
+            .select(enumIdField)
+            .from(enumTable)
+            .where(speciesIdField.eq(speciesId))
+            .fetch(enumIdField.asNonNullable())
             .toSet()
-    val typesToInsert = ecosystemTypes - existingEcosystemTypes
-    val typesToDelete = existingEcosystemTypes - ecosystemTypes
+    val toInsert = values - existing
+    val toDelete = existing - values
 
-    if (typesToDelete.isNotEmpty()) {
+    if (toDelete.isNotEmpty()) {
       dslContext
-          .deleteFrom(SPECIES_ECOSYSTEM_TYPES)
-          .where(SPECIES_ECOSYSTEM_TYPES.SPECIES_ID.eq(speciesId))
-          .and(SPECIES_ECOSYSTEM_TYPES.ECOSYSTEM_TYPE_ID.`in`(typesToDelete))
+          .deleteFrom(enumTable)
+          .where(speciesIdField.eq(speciesId))
+          .and(enumIdField.`in`(toDelete))
           .execute()
     }
 
-    if (typesToInsert.isNotEmpty()) {
+    if (toInsert.isNotEmpty()) {
       dslContext
-          .insertInto(
-              SPECIES_ECOSYSTEM_TYPES,
-              SPECIES_ECOSYSTEM_TYPES.SPECIES_ID,
-              SPECIES_ECOSYSTEM_TYPES.ECOSYSTEM_TYPE_ID)
-          .valuesOfRows(typesToInsert.map { DSL.row(speciesId, it) })
+          .insertInto(enumTable, speciesIdField, enumIdField)
+          .valuesOfRows(toInsert.map { DSL.row(speciesId, it) })
           .execute()
     }
   }
 
-  private fun updateGrowthForms(speciesId: SpeciesId, growthForms: Set<GrowthForm>) {
-    val existingGrowthForms =
-        dslContext
-            .select(SPECIES_GROWTH_FORMS.GROWTH_FORM_ID)
-            .from(SPECIES_GROWTH_FORMS)
-            .where(SPECIES_GROWTH_FORMS.SPECIES_ID.eq(speciesId))
-            .fetch(SPECIES_GROWTH_FORMS.GROWTH_FORM_ID.asNonNullable())
-            .toSet()
-    val formsToInsert = growthForms - existingGrowthForms
-    val formsToDelete = existingGrowthForms - growthForms
-
-    if (formsToDelete.isNotEmpty()) {
-      dslContext
-          .deleteFrom(SPECIES_GROWTH_FORMS)
-          .where(SPECIES_GROWTH_FORMS.SPECIES_ID.eq(speciesId))
-          .and(SPECIES_GROWTH_FORMS.GROWTH_FORM_ID.`in`(formsToDelete))
-          .execute()
-    }
-
-    if (formsToInsert.isNotEmpty()) {
-      dslContext
-          .insertInto(
-              SPECIES_GROWTH_FORMS,
-              SPECIES_GROWTH_FORMS.SPECIES_ID,
-              SPECIES_GROWTH_FORMS.GROWTH_FORM_ID)
-          .valuesOfRows(formsToInsert.map { DSL.row(speciesId, it) })
-          .execute()
-    }
-  }
+  private fun updateEcosystemTypes(speciesId: SpeciesId, ecosystemTypes: Set<EcosystemType>) =
+      updateSet(
+          speciesId,
+          ecosystemTypes,
+          SPECIES_ECOSYSTEM_TYPES,
+          SPECIES_ECOSYSTEM_TYPES.ECOSYSTEM_TYPE_ID,
+          SPECIES_ECOSYSTEM_TYPES.SPECIES_ID)
```

Comment:
Totally bikeshedding, but I notice the first two parameters are species ID then enum, whereas the last two parameters are in the opposite order.

Reply:
Ah yes, I will put it in alphabetical order with param names